### ISSUE TITLE
indentation issue for rust mode

### ISSUE THREAD

**Issue by sunwayforever** [joaotavora/yasnippet#782]
the snippet:
```
fn ${1:name} (${2:args}) -> ${3:()} {
   $0
}
```
before expand:
```
fn main() {
    f
}
```
after expands :
```
fn main() {
    fn name() -> Type {
    
}
}
```
i think the `->` in the snippet cause this bug.

**Comment by npostavs** [comment]
Hmm, it doesn't happen when I step through `yas--indent` with edebug, how annoying.

**Comment by npostavs** [comment]
After tracing a bunch of functions, the problems seems to be that the `syntax-ppss` cache is messed up somehow. `rust-mode` uses `syntax-ppss` in its indentation code.

I guess this code in `yas-expand-snippet` might be relevant:
```elisp
             (let ((yas--inhibit-overlay-hooks t)
                   ;; Avoid major-mode's syntax propertizing function,
                   ;; since we mess with the syntax-table and also
                   ;; insert things that are not valid in the
                   ;; major-mode language syntax anyway.
                   (syntax-propertize-function nil))
               (insert content)
               (setq snippet
                     (yas--snippet-create expand-env start (point))))
             ;; Invalidate any syntax-propertizing done while `syntax-propertize-function' was nil
             (syntax-ppss-flush-cache start))
```

---
```elisp
;; some helper code for tracing
(defconst cursor-string "┊") 
(defun buffer-substring-with-cursor (&optional line)
  (let ((beg (if line (line-beginning-position) (point-min)))
	(end (if line (line-end-position) (point-max))))
    (concat (buffer-substring-no-properties beg (point))
	    cursor-string
	    (buffer-substring-no-properties (point) end))))
```

**Comment by npostavs** [comment]
I think this is a rust-mode bug, I posted analysis and a workaround here: https://github.com/rust-lang/rust-mode/issues/195#issuecomment-285915798

Not sure if we should add the workaround to yasnippet.

**Comment by npostavs** [comment]
rust-mode is now fixed, along with the yasnippet-side fix in #788, this should no longer be a problem.
